lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1308 @@
+    });
+  }
+
+  void _handleConnectionStop(dynamic event) {
+    deviceStatus.element.text = '';
+
+    scriptsView.clearScripts();
+    _scriptCountDiv.text = '0';
+
+    debuggerState.switchToIsolate(null);
+    debuggerState.dispose();
+  }
+
+  @override
+  HelpInfo get helpInfo => null;
+
+  void _populateFromIsolate(Isolate isolate) async {
+    final ScriptList scriptList =
+        await serviceManager.service.getScripts(isolate.id);
+    final List<ScriptRef> scripts = scriptList.scripts.toList();
+
+    debuggerState.scripts = scripts;
+
+    debuggerState.setRootLib(isolate.rootLib);
+    debuggerState.updateFrom(isolate);
+
+    final bool isRunning = isolate.pauseEvent == null ||
+        isolate.pauseEvent.kind == EventKind.kResume;
+
+    scriptsView.showScripts(
+      scripts,
+      debuggerState.rootLib.uri,
+      debuggerState.commonScriptPrefix,
+      selectRootScript: isRunning,
+    );
+    _scriptCountDiv.text = scripts.length.toString();
+  }
+
+  void _displaySource(Script script) {
+    if (script == null) {
+      _sourcePathDiv.setInnerHtml('&nbsp;');
+      sourceEditor.displayScript(script);
+    } else {
+      _sourcePathDiv.text = script.uri;
+      sourceEditor.displayScript(script);
+    }
+  }
+
+  Future<String> _retrieveFullStringValue(InstanceRef stringRef) async {
+    if (stringRef.valueAsStringIsTruncated != true) {
+      return stringRef.valueAsString;
+    }
+
+    final dynamic result = await serviceManager.service.getObject(
+        debuggerState.isolateRef.id, stringRef.id,
+        offset: 0, count: stringRef.length);
+    if (result is Instance) {
+      final Instance obj = result;
+      return obj.valueAsString;
+    } else {
+      return '${stringRef.valueAsString}...';
+    }
+  }
+}
+
+class DebuggerState {
+  DebuggerState();
+
+  VmService service;
+
+  StreamSubscription<Event> _debugSubscription;
+
+  IsolateRef isolateRef;
+  List<ScriptRef> scripts;
+
+  final Map<String, Script> _scriptCache = <String, Script>{};
+
+  final BehaviorSubject<bool> _paused =
+      new BehaviorSubject<bool>(seedValue: false);
+  final BehaviorSubject<bool> _supportsStepping =
+      new BehaviorSubject<bool>(seedValue: false);
+
+  Event _lastEvent;
+
+  final BehaviorSubject<List<Breakpoint>> _breakpoints =
+      new BehaviorSubject<List<Breakpoint>>(seedValue: <Breakpoint>[]);
+
+  final BehaviorSubject<String> _exceptionPauseMode = new BehaviorSubject();
+
+  InstanceRef _reportedException;
+
+  bool get isPaused => _paused.value;
+
+  Stream<bool> get onPausedChanged => _paused;
+
+  Stream<bool> get onSupportsStepping =>
+      new Observable<bool>.concat(<Stream<bool>>[_paused, _supportsStepping]);
+
+  Stream<List<Breakpoint>> get onBreakpointsChanged => _breakpoints;
+
+  Stream<String> get onExceptionPauseModeChanged => _exceptionPauseMode;
+
+  List<Breakpoint> get breakpoints => _breakpoints.value;
+
+  void setVmService(VmService service) {
```

Comment:
how would the user of this API know to use this instead of the setter directly?

How bout making the field private with a public set and get to attach the extra behavior here.